model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java

```diff
@@ -0,0 +1,849 @@
+                .map(entity -> userEntityToAdapterFunc(realm).apply(entity));
+    }
+
+    @Override
+    public long getOfflineSessionsCount(RealmModel realm, ClientModel client) {
+        ModelCriteriaBuilder<UserSessionModel> mcb = realmAndOfflineCriteriaBuilder(realm, true)
+                .compare(UserSessionModel.SearchableFields.CLIENT_ID, ModelCriteriaBuilder.Operator.EQ, client.getId());
+
+        LOG.tracef("getOfflineSessionsCount(%s, %s)%s", realm, client, getShortStackTrace());
+
+        return userSessionTx.getUpdatedNotRemoved(mcb).count();
+    }
+
+    @Override
+    public Stream<UserSessionModel> getOfflineUserSessionsStream(RealmModel realm, ClientModel client,
+                                                                 Integer firstResult, Integer maxResults) {
+        ModelCriteriaBuilder<UserSessionModel> mcb = realmAndOfflineCriteriaBuilder(realm, true)
+                .compare(UserSessionModel.SearchableFields.CLIENT_ID, ModelCriteriaBuilder.Operator.EQ, client.getId());
+
+        LOG.tracef("getOfflineUserSessionsStream(%s, %s, %s, %s)%s", realm, client, firstResult, maxResults, getShortStackTrace());
+
+        return paginatedStream(userSessionTx.getUpdatedNotRemoved(mcb)
+                .map(entity -> userEntityToAdapterFunc(realm).apply(entity)), firstResult, maxResults);
+    }
+
+    @Override
+    public void importUserSessions(Collection<UserSessionModel> persistentUserSessions, boolean offline) {
+        if (persistentUserSessions == null || persistentUserSessions.isEmpty()) {
+            return;
+        }
+
+        persistentUserSessions.stream()
+            .map(pus -> {
+                MapUserSessionEntity userSessionEntity = new MapUserSessionEntity(UUID.randomUUID(), pus.getRealm(), pus.getUser(),
+                        pus.getLoginUsername(), pus.getIpAddress(), pus.getAuthMethod(),
+                        pus.isRememberMe(), pus.getBrokerSessionId(), pus.getBrokerUserId(), offline);
+
+                for (Map.Entry<String, AuthenticatedClientSessionModel> entry : pus.getAuthenticatedClientSessions().entrySet()) {
+                    MapAuthenticatedClientSessionEntity clientSession = toAuthenticatedClientSessionEntity(entry.getValue(), offline);
+
+                    // Update timestamp to same value as userSession. LastSessionRefresh of userSession from DB will have correct value
+                    clientSession.setTimestamp(userSessionEntity.getLastSessionRefresh());
+
+                    userSessionEntity.addAuthenticatedClientSession(entry.getKey(), clientSession.getId());
+
+                    clientSessionTx.create(clientSession.getId(), clientSession);
+                }
+
+                return userSessionEntity;
+            })
+            .forEach(use -> userSessionTx.create(use.getId(), use));
+    }
+
+    @Override
+    public void close() {
+
+    }
+
+    public void removeLocalOfflineUserSessions(RealmModel realm) {
```

Comment:
This doesn't look very pretty. What is the difference between this and `removeUserSessions(RealmModel realm)` can we use that function instead? 

If this is only for testsuite this can be also solved by some custom testsuite provider that is used in the corresponding test.

Reply:
The difference is that one is dealing with online sessions and the other one with offline session.
Exactly, it's used just in tests and I followed the same pattern as was used for Infinispan provider.

Reply:
Well, that is exactly what I don't really like about it. Having a public method in Keycloak server source code, that is supposed to be used only in tests. But maybe I am missing something, so if you are sure this is necessary I won't make any trouble.

Just some idea, in the particular test where this is used, it seems it could be replaced by (but again, maybe I am missing something, you have much more insight into it):
```
Stream.of(session.users().getUserByUsername(realm, "user1"), session.users().getUserByUsername(realm, "user2"))
                        .flatMap(model -> provider.getUserSessionsStream(realm, model))
                        .filter(UserSessionModel::isOffline)
                        .forEach(uSession -> provider.removeUserSession(realm, uSession));
```

Reply:
This solution looks good! I used it instead of the public method.